RecoEgamma/EgammaTools/plugins/DRNCorrectionProducerT.cc

```diff
@@ -0,0 +1,447 @@
+      vdataxECAL.push_back(rescale(pos.y(), XY_MIN, XY_RANGE));
+      vdataxECAL.push_back(rescale(pos.z(), Z_MIN, Z_RANGE));
+      vdataxECAL.push_back(rescale(hit->energy() * detitr.second, ECAL_MIN, ECAL_RANGE));
+      vdataxECAL.push_back(rescale(ped->find(detitr.first)->rms(1), NOISE_MIN, NOISE_RANGE));
+
+      //fill fECAL
+      int64_t flagVal = 0;
+      if (hit->checkFlag(EcalRecHit::kGood))
+        flagVal += 1;
+      if (hit->checkFlag(EcalRecHit::kOutOfTime))
+        flagVal += 2;
+      if (hit->checkFlag(EcalRecHit::kPoorCalib))
+        flagVal += 4;
+
+      vdatafECAL.push_back(flagVal);
+
+      //fill gain
+      int64_t gainVal = 0;
+      if (hit->checkFlag(EcalRecHit::kHasSwitchToGain6))
+        gainVal = 1;
+      else if (hit->checkFlag(EcalRecHit::kHasSwitchToGain1))
+        gainVal = 0;
+      else
+        gainVal = 2;
+
+      vdataGainECAL.push_back(gainVal);
+
+      //fill batch number
+      vdataBatchECAL.push_back(partNum);
+    }  //end iterate over ECAL hits
+
+    //iterate over ES clusters...
+    for (auto iES = sc->preshowerClustersBegin(); iES != sc->preshowerClustersEnd(); ++iES) {
+      for (const auto& ESitr : (*iES)->hitsAndFractions()) {  //iterate over ES hits
+        hit = recHitsES->find(ESitr.first);
+        geom = ecalESGeom->getGeometry(ESitr.first);
+        auto& pos = geom->getPosition();
+
+        //fill xES
+        vdataxES.push_back(rescale(pos.x(), XY_MIN, XY_RANGE));
+        vdataxES.push_back(rescale(pos.y(), XY_MIN, XY_RANGE));
+        vdataxES.push_back(rescale(pos.z(), Z_MIN, Z_RANGE));
+        vdataxES.push_back(rescale(hit->energy(), ES_MIN, ES_RANGE));
+
+        //fill fES
+        int64_t flagVal = 0;
+        if (hit->checkFlag(EcalRecHit::kESGood))
+          flagVal += 1;
+
+        vdatafES.push_back(flagVal);
+
+        //fill batchES
+        vdataBatchES.push_back(partNum);
+      }  //end iterate over ES hits
+    }    //end iterate over ES clusters
+
+    //fill gx
+    vdataGx.push_back(rescale(rho, RHO_MIN, RHO_RANGE));
+    vdataGx.push_back(rescale(part.hadronicOverEm(), HOE_MIN, HOE_RANGE));
+
+    //increment particle number
+    ++partNum;
+  }  // end iterate over particles
+
+  /*
+   * Convert input tensors to server data format
+   */
+  inputxECAL.toServer(dataxECAL);
+  inputfECAL.toServer(datafECAL);
+  inputGainECAL.toServer(dataGainECAL);
+  inputBatchECAL.toServer(dataBatchECAL);
+
+  inputGx.toServer(dataGx);
+
+  inputxES.toServer(dataxES);
+  inputfES.toServer(datafES);
+  inputBatchES.toServer(dataBatchES);
+}
+
+template <typename T>
+void DRNCorrectionProducerT<T>::produce(edm::Event& iEvent, const edm::EventSetup& iSetup, Output const& iOutput) {
+  particles_ = iEvent.getHandle(particleToken_);
+
+  std::vector<std::pair<float, float>> corrections;
+  corrections.reserve(nPart_);
+
+  //if there are no particles, the fromServer() call will fail
+  //but we can just put() an empty valueMap
+  if (nPart_) {
```

Comment:
I may be confused here about the async logic. 

The class member `nPart_` is filled in `acquire()` for a particular event. Won't storing the event state in the class mess up the async logic?
 
For example, you can imagine the following async order of execution 
```
acquire(event1);
acquire(event2);
acquire(event3);
```
while for produce you might conceivably have
```
produce(event1);
produce(event3);
produce(event2);
```

does it make sense to store state in the class members?